pkg/maps/egressmap/egress.go

```diff
@@ -4,79 +4,72 @@
 package egressmap
 
 import (
+	"errors"
 	"fmt"
 	"net"
-	"unsafe"
 
-	"github.com/cilium/cilium/pkg/bpf"
-	"github.com/cilium/cilium/pkg/types"
-)
+	"github.com/cilium/cilium/pkg/ebpf"
+	"github.com/cilium/cilium/pkg/logging"
+	"github.com/cilium/cilium/pkg/logging/logfields"
 
-const (
-	MapName    = "cilium_egress_v4"
-	MaxEntries = 16384
+	"github.com/sirupsen/logrus"
 )
 
-// +k8s:deepcopy-gen=true
-// +k8s:deepcopy-gen:interfaces=github.com/cilium/cilium/pkg/bpf.MapKey
-type Key4 struct {
-	// PrefixLen is full 32 bits of SourceIP + DestCIDR's mask bits
-	PrefixLen uint32
+var log = logging.DefaultLogger.WithField(logfields.LogSubsys, "map-egress")
 
-	SourceIP types.IPv4
-	DestCIDR types.IPv4
-}
+const (
+	PolicyMapName = "cilium_egress_gw_policy_v4"
 
-func (k *Key4) GetKeyPtr() unsafe.Pointer { return unsafe.Pointer(k) }
-func (k *Key4) NewValue() bpf.MapValue    { return &EgressInfo4{} }
-func (k *Key4) String() string {
-	return fmt.Sprintf("%s %s/%d", k.SourceIP, k.DestCIDR, k.PrefixLen-getStaticPrefixBits())
-}
+	MaxPolicyEntries = 1 << 14
+)
+
+var (
+	EgressPolicyMap *egressPolicyMap
+)
 
-func getStaticPrefixBits() uint32 {
-	staticMatchSize := unsafe.Sizeof(Key4{})
-	staticMatchSize -= unsafe.Sizeof(Key4{}.PrefixLen)
-	staticMatchSize -= unsafe.Sizeof(Key4{}.DestCIDR)
-	return uint32(staticMatchSize) * 8
+// ApplyEgressPolicy adds a new entry to the egress policy map.
+// If a policy with the same key already exists, it will get replaced.
+func ApplyEgressPolicy(sourceIP net.IP, destCIDR net.IPNet, egressIP, gatewayIP net.IP) error {
+	logger := log.WithFields(logrus.Fields{
+		logfields.SourceIP:        sourceIP,
+		logfields.DestinationCIDR: destCIDR,
+		logfields.EgressIP:        egressIP,
+		logfields.GatewayIP:       gatewayIP,
+	})
+
+	if err := EgressPolicyMap.Update(sourceIP, destCIDR, egressIP, gatewayIP); err != nil {
+		return fmt.Errorf("cannot apply egress policy: %w", err)
+	}
+
+	logger.Info("Applied egress policy")
```

Comment:
Hm, wonder if it'd be better to log both the error and success case in the same place? When I looked at this function I immediately wondered if we log the returned error somewhere (yep, in OnAddEgressPolicy) and if we log it with enough details.

Reply:
Yep, makes sense. I moved all the logging into the manager and also got rid of `egressmap.{ApplyEgressPolicy, RemoveEgressPolicyMap}` since those methods where just wrapping a single function call (`egressmap.EgressPolicyMap.{Update, Delete}`)